### PR TITLE
added iw

### PR DESCRIPTION
**Pull request by lhleohuang**

### CODE REVIEW

File: efficient_reasoning/grpo_trainer.py [hunk 1]
```diff
@@ -609,8 +623,8 @@ def data_collator(features):  # No data collation is needed in GRPO
         if self.ref_model is not None:
             if self.is_deepspeed_enabled:
                 self.ref_model = prepare_deepspeed(self.ref_model, self.accelerator)
-        else:
-            self.ref_model = self.accelerator.prepare_model(self.ref_model, evaluation_mode=True)
+            else:
```
Comment (lhleohuang) [inline efficient_reasoning/grpo_trainer.py:627]: I checked source code (https://github.com/huggingface/trl/blob/822653824bf084bc6c042cf0e759f86187c92569/trl/trainer/grpo_trainer.py), I think mine is correct here. Idk why the identation was different.

File: efficient_reasoning/grpo_trainer.py [hunk 2]
```diff
@@ -871,28 +923,6 @@ def _generate_and_score_completions(
 
         logits_to_keep = completion_ids.size(1)  # we only need to compute the logits for the completion tokens
```
Comment (lhleohuang) [inline efficient_reasoning/grpo_trainer.py:941]: moved to later

File: efficient_reasoning/grpo_trainer.py [hunk 3]
```diff
@@ -985,6 +1015,21 @@ def _generate_and_score_completions(
         self._metrics[mode]["completions/min_length"].append(agg_completion_mask.float().min().item())
         self._metrics[mode]["completions/max_length"].append(agg_completion_mask.float().max().item())
```
Comment (lhleohuang) [inline efficient_reasoning/grpo_trainer.py:1033]: extra logging - nothing to do with iw

File: efficient_reasoning/grpo_trainer.py [hunk 4]
```diff
@@ -1021,12 +1066,117 @@ def _generate_and_score_completions(
         for i, name in enumerate(reward_func_names):
             self._textual_logs["rewards"][name].extend(rewards_per_func[:, i].tolist())
```
Comment (lhleohuang) [inline efficient_reasoning/grpo_trainer.py:1084]: in this version, gradient_filtering is done in this method rather than in compute_loss

nothing to do with iw

File: efficient_reasoning/grpo_trainer.py [hunk 5]
```diff
@@ -1021,12 +1066,117 @@ def _generate_and_score_completions(
...
+                prompt_completion_ids = prompt_completion_ids[valid]
+                attention_mask = attention_mask[valid]
+                advantages_filtered = advantages[valid]
+                completion_ids = completion_ids[valid]
+                logits_to_keep = completion_ids.size(1)
+                completion_mask = completion_mask[valid]
+                prompt_ids = prompt_ids[valid]
+                prompt_mask = prompt_mask[valid]
+
+
```
Comment (lhleohuang) [inline efficient_reasoning/grpo_trainer.py:1112]: the following if with self.iw is key

File: efficient_reasoning/grpo_trainer.py [hunk 6]
```diff
@@ -1021,12 +1066,117 @@ def _generate_and_score_completions(
...
+            self._metrics[mode]["advantages/std"].append(all_advantages.float().abs().std().item())
+            self._metrics[mode]["advantages/min"].append(all_advantages.float().abs().min().item())
+            self._metrics[mode]["advantages/max"].append(all_advantages.float().abs().max().item())
+
         return {
             "prompt_ids": prompt_ids,
             "prompt_mask": prompt_mask,
             "completion_ids": completion_ids,
             "completion_mask": completion_mask,
             "advantages": advantages,
```
Comment (lhleohuang) [inline efficient_reasoning/grpo_trainer.py:1194]: the original rewards need to be returned for importance weighted advantage
